File: quizzes/quizzesV2/quizzes6/05-problem.ts
export {};

type Entity = { id: number };

type User = Entity & { name: string };

type Post = Entity & {
  user_id: User["id"];
  title: string;
};

type PostData = {
  users: User[];
  posts: Post[];
};

type PostItem = Omit<Post, "user_id">;

type UserPosts = User & { posts: PostItem[] };

const data: PostData = {
  users: [
    { id: 1, name: "Alice" },
    { id: 2, name: "Bob" },
  ],
  posts: [
    { id: 101, user_id: 1, title: "Hello" },
    { id: 102, user_id: 1, title: "World" },
    { id: 103, user_id: 2, title: "JS Rocks" },
  ],
};

const { users, posts } = data;

const getUsersPostsWithObj = (): UserPosts[] => {
  const groupedPosts: Partial<Record<number, Post[]>> = Object.groupBy(
    posts,
    (post) => post.user_id,
  );

  return users.map((user) => {
    const userPosts: Post[] = groupedPosts[user.id] ?? [];

    return {
      ...user,
      posts: userPosts.map(({ id, title }) => ({ id, title })),
    };
  });
};

const getUsersPostsWithLoop = (): UserPosts[] => {
  const postsByUserId = new Map<User["id"], PostItem[]>();

  for (const { user_id, id, title } of posts) {
    const group = postsByUserId.get(user_id) ?? [];

    if (!postsByUserId.has(user_id)) {
      postsByUserId.set(user_id, group);
    }

    group.push({ id, title });
  }

  return users.map((user) => {
    return {
      ...user,
      posts: postsByUserId.get(user.id) ?? [],
    };
  });
};

console.log(JSON.stringify({ data: getUsersPostsWithObj() }, null, 2));
console.log(JSON.stringify({ data: getUsersPostsWithLoop() }, null, 2));

File: bins3/05-problem.rb
# frozen_string_literal: true

require "json"

data = {
  "users": [
    { "id": 1, "name": "Alice" },
    { "id": 2, "name": "Bob" }
  ],
  "posts": [
    { "id": 101, "user_id": 1, "title": "Hello" },
    { "id": 102, "user_id": 1, "title": "World" },
    { "id": 103, "user_id": 2, "title": "JS Rocks" }
  ]
}

users = data[:users]
posts = data[:posts]

def get_users_posts_with_group(users, posts)
  grouped_posts = posts.group_by { |post| post[:user_id] }

  users.map do |user|
    user_posts = (grouped_posts[user[:id]] || []).map do |p|
      {
        id: p[:id],
        title: p[:title]
      }
    end

    user.merge(posts: user_posts)
  end
end

def get_users_posts_with_loop(users, posts)
  posts_by_user_id = {}

  posts.each do |post|
    user_id = post[:user_id]
    posts_by_user_id[user_id] ||= []
    posts_by_user_id[user_id] << { id: post[:id], title: post[:title] }
  end

  users.map do |user|
    user.merge(posts: posts_by_user_id[user[:id]] || [])
  end
end

puts JSON.pretty_generate({ data: get_users_posts_with_group(users, posts) })
puts JSON.pretty_generate({ data: get_users_posts_with_loop(users, posts) })
